File: src/tools/context.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import {
  buildContext,
  checkpoint,
  getWorkingState,
  setWorkingState,
  updateProjectContextSummary,
} from "../db.js";

export function registerContextTools(server: McpServer) {
  server.tool(
    "set_working_state",
    `Stores the active working state for one session.
Use during work whenever focus, open threads, or pinned memory changes. This makes build_context prioritize the right entities for this session.`,
    {
      session_id: z.string().describe("Session ID"),
      focus: z.string().describe("Current work focus"),
      open_threads: z.array(z.string()).describe("Open topics or unresolved threads"),
      pinned_ids: z
        .array(z.string())
        .describe("Pinned IDs. May use raw id or type:id, for example code_entity:abc"),
    },
    async ({ session_id, focus, open_threads, pinned_ids }) => {
      setWorkingState(session_id, focus, open_threads, pinned_ids);
      return { content: [{ type: "text", text: "Working state updated." }] };
    },
  );

  server.tool(
    "get_working_state",
    "Returns active focus, open threads, and pinned memory for one session.",
    {
      session_id: z.string().describe("Session ID"),
    },
    async ({ session_id }) => {
      const result = getWorkingState(session_id);
      return { content: [{ type: "text", text: JSON.stringify(result, null, 2) }] };
    },
  );

  server.tool(
    "update_project_context_summary",
    `Updates the compact project-level context summary.
Use when stable project state changes: current capabilities, architecture, constraints, important pending work.
The agent/model writes this summary. The MCP stores it.`,
    {
      project_id: z.string().describe("Project ID"),
      context_summary: z.string().describe("Compact stable project context summary"),
    },
    async ({ project_id, context_summary }) => {
      updateProjectContextSummary(project_id, context_summary);
      return { content: [{ type: "text", text: "Project context summary updated." }] };
    },
  );

  server.tool(
    "build_context",
    `Builds a token-aware context bundle from persistent DB memory.
Use after identifying the project and before answering/working on project-specific tasks. This is the main source-of-truth load path.
Includes active notes, decisions, artifacts, and code_entities. Prioritizes this session's pinned_ids, then importance, then recency. Respects token_budget with an internal margin.
If session_id is omitted, uses global pins only for backward compatibility.`,
    {
      project_id: z.string().describe("Project ID"),
      token_budget: z.number().describe("Maximum token budget for returned context"),
      session_id: z
        .string()
        .optional()
        .describe("Session ID used to prioritize only that session's pinned_ids"),
    },
    async ({ project_id, token_budget, session_id }) => {
      const result = buildContext(project_id, token_budget, session_id);
      return { content: [{ type: "text", text: JSON.stringify(result, null, 2) }] };
    },
  );

  server.tool(
    "checkpoint",
    `Consolidates a session or work stage into persistent memory.
Use at the end of a meaningful stage, and also during long sessions after major changes.
The agent/model must provide dense summaries when possible. The MCP does not call an LLM.
session_summary: compact session index (~300 tokens max): goal, tools/functions touched, decisions, changes, final state.
context_summary: stable project state: current capabilities, architecture/constraints, pending work. No transcript or narrative.
If context_summary is omitted, MCP generates a mechanical fallback from stored DB context.`,
    {
      session_id: z.string().describe("Session ID to consolidate"),
      project_id: z.string().describe("Associated project ID"),
      session_summary: z
        .string()
        .optional()
        .describe("Dense session summary (~300 tokens max); replaces sessions.summary if provided"),
      context_summary: z
        .string()
        .optional()
        .describe("Stable project summary; if omitted, a mechanical fallback is generated"),
      token_budget: z
        .number()
        .optional()
        .describe("Budget for the associated build_context call (default: 4000)"),
    },
    async ({
      session_id,
      project_id,
      session_summary,
      context_summary,
      token_budget,
    }) => {
      const result = checkpoint(session_id, project_id, {
        session_summary,
        context_summary,
        token_budget,
      });
      return { content: [{ type: "text", text: JSON.stringify(result, null, 2) }] };
    },
  );
}

File: rust/src/tools/context.rs
//! Context tools. Mirrors `src/tools/context.ts`.

use rmcp::{
    handler::server::wrapper::Parameters,
    model::CallToolResult,
    schemars, tool, tool_router, ErrorData,
};
use serde::Deserialize;

use crate::repo::{context as repo_context, projects, session_focus, working_state};
use crate::summary::{ContextSummary, SessionSummary};

use super::{json_result, repo_error, MemoryService};

#[derive(Debug, Deserialize, schemars::JsonSchema)]
pub struct SetWorkingStateArgs {
    pub session_id: String,
    pub focus: String,
    /// Pinned IDs. May use raw id or type:id, for example code_entity:abc.
    pub pinned_ids: Vec<String>,
}

#[derive(Debug, Deserialize, schemars::JsonSchema)]
pub struct GetWorkingStateArgs {
    pub session_id: String,
}

#[derive(Debug, Deserialize, schemars::JsonSchema)]
pub struct SetFocusArgs {
    pub session_id: String,
    pub project_id: String,
    /// Non-empty description of what the session is working on right now.
    pub focus: String,
}

#[derive(Debug, Deserialize, schemars::JsonSchema)]
pub struct GetFocusArgs {
    pub session_id: String,
}

#[derive(Debug, Deserialize, schemars::JsonSchema)]
pub struct GetLatestFocusForProjectArgs {
    pub project_id: String,
}

#[derive(Debug, Deserialize, schemars::JsonSchema)]
pub struct UpdateProjectContextSummaryArgs {
    pub project_id: String,
    pub context_summary: ContextSummary,
}

#[derive(Debug, Deserialize, schemars::JsonSchema)]
pub struct BuildContextArgs {
    pub project_id: String,
    #[serde(deserialize_with = "super::flex_int::deserialize")]
    pub token_budget: i64,
    #[serde(default)]
    pub session_id: Option<String>,
    /// Tokenizer model for accurate budget accounting. Supported: 'anthropic:claude', \
    /// 'openai:o200k_base', 'openai:cl100k_base'.
    #[serde(default)]
    pub tokenizer_model: Option<String>,
}

#[derive(Debug, Deserialize, schemars::JsonSchema)]
pub struct CheckpointArgs {
    pub session_id: String,
    pub project_id: String,
    #[serde(default)]
    pub session_summary: Option<SessionSummary>,
    #[serde(default)]
    pub context_summary: Option<ContextSummary>,
    /// Budget for the associated build_context call (default: 4000).
    #[serde(default, deserialize_with = "super::flex_int::opt::deserialize")]
    pub token_budget: Option<i64>,
}

#[tool_router(router = context_router, vis = "pub(crate)")]
impl MemoryService {
    #[tool(
        description = "Stores focus + pinned_ids for one session. `open_threads` was removed — \
            use `open_thread` / `close_thread` for lifecycle-tracked threads instead. Focus set \
            here is mirrored to `session_focus` when `set_focus` is used explicitly."
    )]
    pub async fn set_working_state(
        &self,
        Parameters(args): Parameters<SetWorkingStateArgs>,
    ) -> Result<CallToolResult, ErrorData> {
        working_state::set(&self.db, &args.session_id, &args.focus, &args.pinned_ids)
            .map_err(repo_error)?;
        json_result(&serde_json::json!({ "ok": true }))
    }

    #[tool(
        description = "Upsert the current focus of a session, scoped to a project. Atomic and \
            small — preferred over set_working_state when you only want to change focus. \
            The harness may require this to be set before persistent actions."
    )]
    pub async fn set_focus(
        &self,
        Parameters(args): Parameters<SetFocusArgs>,
    ) -> Result<CallToolResult, ErrorData> {
        let row = session_focus::set(&self.db, &args.session_id, &args.project_id, &args.focus)
            .map_err(repo_error)?;
        // Mirror to working_state so build_context and legacy consumers see it.
        let _ = working_state::set(&self.db, &args.session_id, &args.focus, &[]);
        json_result(&row)
    }

    #[tool(description = "Returns the focus record for one session, or null.")]
    pub async fn get_focus(
        &self,
        Parameters(args): Parameters<GetFocusArgs>,
    ) -> Result<CallToolResult, ErrorData> {
        let row = session_focus::get(&self.db, &args.session_id).map_err(repo_error)?;
        json_result(&row)
    }

    #[tool(
        description = "Returns the most recently updated focus for a project across all its \
            sessions, or null. Cheap read the harness uses to compute `stats.last_focus`."
    )]
    pub async fn get_latest_focus_for_project(
        &self,
        Parameters(args): Parameters<GetLatestFocusForProjectArgs>,
    ) -> Result<CallToolResult, ErrorData> {
        let row = session_focus::get_latest_for_project(&self.db, &args.project_id)
            .map_err(repo_error)?;
        json_result(&row)
    }

    #[tool(description = "Returns focus and pinned memory for one session.")]
    pub async fn get_working_state(
        &self,
        Parameters(args): Parameters<GetWorkingStateArgs>,
    ) -> Result<CallToolResult, ErrorData> {
        let ws = working_state::get(&self.db, &args.session_id).map_err(repo_error)?;
        json_result(&ws)
    }

    #[tool(description = "Updates the STRUCTURED project-level context summary.")]
    pub async fn update_project_context_summary(
        &self,
        Parameters(args): Parameters<UpdateProjectContextSummaryArgs>,
    ) -> Result<CallToolResult, ErrorData> {
        projects::update_context_summary(&self.db, &args.project_id, &args.context_summary)
            .map_err(repo_error)?;
        json_result(&serde_json::json!({ "ok": true }))
    }

    #[tool(
        description = "Builds a token-aware context bundle from persistent DB memory. \
            Use after identifying the project and before answering project-specific tasks."
    )]
    pub async fn build_context(
        &self,
        Parameters(args): Parameters<BuildContextArgs>,
    ) -> Result<CallToolResult, ErrorData> {
        let r = repo_context::build_context(
            &self.db,
            &args.project_id,
            args.token_budget,
            args.session_id.as_deref(),
            args.tokenizer_model.as_deref(),
        )
        .map_err(repo_error)?;
        json_result(&r)
    }

    #[tool(
        description = "Consolidates a session or work stage into persistent memory with STRUCTURED summaries. \
            If context_summary is omitted, MCP merges previous project context with current session pending."
    )]
    pub async fn checkpoint(
        &self,
        Parameters(args): Parameters<CheckpointArgs>,
    ) -> Result<CallToolResult, ErrorData> {
        let options = repo_context::CheckpointOptions {
            session_summary: args.session_summary,
            context_summary: args.context_summary,
            token_budget: args.token_budget,
        };
        let r = repo_context::checkpoint(&self.db, &args.session_id, &args.project_id, &options)
            .map_err(repo_error)?;
        json_result(&r)
    }
}
